internal/integration/provision/provision.go

```diff
@@ -71,3 +103,533 @@ func trimVersion(version string) string {
 	// remove anything extra after semantic version core, `v0.3.2-1-abcd` -> `v0.3.2`
 	return regexp.MustCompile(`(-\d+-g[0-9a-f]+)$`).ReplaceAllString(version, "")
 }
+
+var defaultNameservers = []netip.Addr{netip.MustParseAddr("8.8.8.8"), netip.MustParseAddr("1.1.1.1")}
+
+// BaseSuite provides base features for provision tests.
+type BaseSuite struct {
+	suite.Suite
+	base.TalosSuite
+
+	provisioner provision.Provisioner
+
+	configBundle *bundle.Bundle
+
+	clusterAccess        *access.Adapter
+	controlPlaneEndpoint string
+
+	//nolint:containedctx
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	stateDir string
+	cniDir   string
+}
+
+// SetupSuite ...
+func (suite *BaseSuite) SetupSuite() {
+	// timeout for the whole test
+	suite.ctx, suite.ctxCancel = context.WithTimeout(context.Background(), 50*time.Minute)
+
+	var err error
+
+	suite.provisioner, err = qemu.NewProvisioner(suite.ctx)
+	suite.Require().NoError(err)
+}
+
+// TearDownSuite ...
+func (suite *BaseSuite) TearDownSuite() {
+	if suite.T().Failed() && DefaultSettings.CrashdumpEnabled && suite.Cluster != nil {
+		// for failed tests, produce crash dump for easier debugging,
+		// as cluster is going to be torn down below
+		suite.provisioner.CrashDump(suite.ctx, suite.Cluster, os.Stderr)
+
+		if suite.clusterAccess != nil {
+			suite.clusterAccess.CrashDump(suite.ctx, os.Stderr)
+		}
+	}
+
+	if suite.clusterAccess != nil {
+		suite.Assert().NoError(suite.clusterAccess.Close())
+	}
+
+	if suite.Cluster != nil {
+		suite.Assert().NoError(suite.provisioner.Destroy(suite.ctx, suite.Cluster))
+	}
+
+	suite.ctxCancel()
+
+	if suite.stateDir != "" {
+		suite.Assert().NoError(os.RemoveAll(suite.stateDir))
+	}
+
+	if suite.provisioner != nil {
+		suite.Assert().NoError(suite.provisioner.Close())
+	}
+}
+
+// waitForClusterHealth asserts cluster health after any change.
+func (suite *BaseSuite) waitForClusterHealth() {
+	runs := 1
+
+	singleNodeCluster := len(suite.Cluster.Info().Nodes) == 1
+	if singleNodeCluster {
+		// run health check several times for single node clusters,
```

Comment:
we don;t use self hosted anymore, right?

Reply:
I think it wasn't specifically because of self-hosted, but rather because a single API server might "lie" about the state, as it's all single kubelet/api-server and they might get out of sync when both are restarted

Reply:
and this is moved from another file, I haven't changed it actually.